### ISSUE TITLE
SAML SP metadata from URL collisions

### ISSUE THREAD

**Issue by rfrovarp** [apereo/cas#2810]
This is using CAS 5.1.1. Looking at the commit log for 5.1.2 I don't see it addressed.

The issue is that we can specify the SAML SP metadata and have it loaded by URL. This is preferable than by file on the file system as it will automatically update over time, rather than us having to build that functionality.

The problem is it appears CAS stores that file on the file system by using whatever the downloaded filename is. If multiple SPs end up with the same name, the one that is on the filesystem / still valid in the cache wins, the others get an error stating that they aren't permitted. Since most SPs out there are going to have a filename of "Metadata" or "metadata" this becomes a problem. On Linux, it appears that "Metadata" (from shibd) and "metadata" (from vendor I'm adding) both work and don't collide because of the case sensitive nature of the OS. However, when trying to add another SP from the same vendor, it collides with the existing "metadata" file and which one works seems to be dependent on time and when either were last accessed.

Seems to me that the on disk cache should have some sort of disambiguous marker applied, however I honestly don't know what that would be other than the full metadata url, which is kind of awkward.

**Comment by mmoayyed** [comment]
> Seems to me that the on disk cache should have some sort of disambiguous marker applied, however I honestly don't know what that would be other than the full metadata url, which is kind of awkward.

What if the filename on disk were saved with System's current time in nanoseconds? Something like `something-metadata-1234567890.xml`?

**Comment by dima767** [comment]
How about UUID?

**Comment by mmoayyed** [comment]
Fine idea.

I did review this a bit, and realized that all backups are kept on disk under a `metadata-backups` directory. To make things easier to spot and navigate, it would likely be best if we created separated directories for each SP based on the registered service name (`RegisteredService#getName()`). 

So the overall directory structure for backups would be, as an example:

```
- /etc/cas/saml/metadata-backups/{service-provider1-name}/something-metadata.xml
- /etc/cas/saml/metadata-backups/{service-provider2-name}/something-metadata.xml
```

So that if you wanted to see the backups, you'd be able to spot them easily and things won't step on each other either with name collisions. If somehow by accident two definitions ended up having the same name (which is an error, but whatever), we can check for the existence of the backup, WARN and then use:

```
- /etc/cas/saml/metadata-backups/{service-provider2-name}-UUID/something-metadata.xml
```

Sound good?
